src/middleware.ts

```diff
@@ -259,43 +296,59 @@ export const persist = <S extends State>(
   }
 
   // rehydrate initial state with existing stored state
-  ;(async () => {
-    const postRehydrationCallback = onRehydrateStorage?.(get()) || undefined
-
-    try {
-      const storageValue = await storage.getItem(name)
 
+  // a workaround to solve the issue of not storing rehydrated state in sync storage
+  // the set(state) value would be later overridden with initial state by create()
+  // to avoid this, we merge the state from localStorage into the initial state.
+  // TODO: find a better solution for this
+  let stateFromStorage: any
+  const postRehydrationCallback = onRehydrateStorage?.(get()) || undefined
+  // bind is used to avoid `TypeError: Illegal invocation` error
+  toThenable(storage.getItem.bind(storage))(name)
+    .then((storageValue) => {
       if (storageValue) {
-        const deserializedStorageValue = await deserialize(storageValue)
-
-        // if versions mismatch, run migration
+        return deserialize(storageValue)
+      }
+    })
+    .then((deserializedStorageValue) => {
+      if (deserializedStorageValue) {
         if (deserializedStorageValue.version !== version) {
-          const migratedState = await migrate?.(
-            deserializedStorageValue.state,
-            deserializedStorageValue.version
-          )
-          if (migratedState) {
-            set(migratedState)
-            await setItem()
+          if (migrate) {
+            stateFromStorage = migrate(
```

Comment:
This is wrong. `migrate` can return a promise. `toThenable(migrate)`

Reply:
Ou, right, I wasn't aware. Then we need to also adapt [this](https://github.com/pmndrs/zustand/blob/master/tests/persist.test.tsx#L154) to use a `async` migrate so that both versions are tested properly. Will do.

Reply:
so stateFromStorage should be inside `L239` `then` handler. right?
I think there is something wrong with `types` the 'handlers' input values are never `promise`. but it detects it as a promise.

Reply:
@barbogast I just move it to the next `then()` I think that would solve the issue.